ci/scripts/utils/rocoto_statcount.py

```diff
@@ -0,0 +1,104 @@
+        If all cycles are not done, then rocoto_state is Running.
+        If the check_stalled is used then rocotorun is then issued and
+        rocotostat is run again and if all jobs do not advanced, then
+        rocoto_state is Stalled and the script exits with -1.
+        """
+
+    parser = ArgumentParser(description=description,
+                            formatter_class=ArgumentDefaultsHelpFormatter)
+
+    parser.add_argument('-w', help='workflow_document', type=str)
+    parser.add_argument('-d', help='database_file', type=str)
+    parser.add_argument('--check_stalled', help='check if any jobs do not advance (stalled)', action='store_true', required=False)
+
+    args = parser.parse_args()
+
+    return args
+
+def rocoto_statcount():
+    """
+    Run rocotostat and process its output.
+    """
+
+    args = input_args()
+
+    rocotostat = which("rocotostat")
+    if not rocotostat:
+        logger.exception("rocotostat not found in PATH")
+        sys.exit(-1)
+
+    xml_file_path = os.path.abspath(args.w)
+    db_file_path = os.path.abspath(args.d)
+
+    rocotostat_all = which("rocotostat")
+    rocotostat.add_default_arg(['-w',xml_file_path,'-d',db_file_path,'-s'])
+    rocotostat_all.add_default_arg(['-w',xml_file_path,'-d',db_file_path,'-a'])
+
+    rocotostat_output = rocotostat(output=str)
+    rocotostat_output = rocotostat_output.splitlines()[1:]
+    rocotostat_output = [line.split()[0:2] for line in rocotostat_output]
+
+    rocotostat_output_all = rocotostat_all(output=str)
+    rocotostat_output_all = rocotostat_output_all.splitlines()[1:]
+    rocotostat_output_all = [line.split()[0:4] for line in rocotostat_output_all]
+    rocotostat_output_all = [line for line in rocotostat_output_all if len(line) != 1]
+
+    rocoto_status = {
+       'Cycles' : len(rocotostat_output),
+       'Done_Cycles' :  sum([ sublist.count('Done') for sublist in rocotostat_output ])
+    }
+    
+    status_cases = [ 'SUCCEEDED', 'FAIL', 'DEAD', 'RUNNING', 'PENDING', 'QUEUED']
+    for case in status_cases:
+        rocoto_status[case] = sum([ sublist.count(case) for sublist in rocotostat_output_all ])
+
+    return rocoto_status
+
+if __name__ == '__main__':
+
+    args = input_args()
+
+    rocoto_status = rocoto_statcount()
+    for status in rocoto_status:
+        print(f'Number of {status} : {rocoto_status[status]}')
+    rocoto_state = 'Running'
+    if rocoto_status['Cycles'] == rocoto_status['Done_Cycles']:
+        rocoto_state = 'Done'
+
+    if args.check_stalled:
+        if rocoto_state != 'Done':
+            rocoto_run = which("rocotorun")
+            rocoto_run.add_default_arg(['-w',args.w,'-d',args.d])
+            rocoto_run()
+            rocoto_status2 = rocoto_statcount()
+            if rocoto_status2 == rocoto_status:
+                rocoto_state = 'Stalled'
+                print(f'Rocoto State : {rocoto_state}')
+                sys.exit(-1)
+            else:
+                rocoto_state = 'Running'
```

Comment:
I don't think this is robust. It's not guaranteed the statuses to change on a `rocotorun` (particularly if a long job is running). As we discussed before, what we should be looking for is the number of successful or pending/queued jobs. You don't even need to call `rocotorun` the second time or have this extra `check_stalled` option.
```suggestion
    if rocoto_status['Cycles'] == rocoto_status['Done_Cycles']:
        rocoto_state = 'Done'
    elif rocoto_status['RUNNING'] + rocoto_status['SUBMITTING'] + rocoto_status['QUEUED'] == 0:
        rocoto_state = 'Stalled'
    else:
        rocoto_state = 'Running'
```

Reply:
`elif any(x in rocoto_status for x in ['RUNNING', 'SUBMITTING', 'QUEUED'])` should do for line 90.

Reply:
Oh right Walter of course, not not advancing doesn't mean the all the status states don't change, its when after  running **rocotorun** there are none that are RUNNING, SUBMITTING, or QUEUED.  Really simplifies things indeed.